### PR TITLE
Fix/instance varcreation

### PR DESCRIPTION
**Pull request by therealplato**
This PR adds `vars` validation for instance creation and setting vars for an existing instance. Both will fail if you pass in a var that is not present in the instance's playbook.

Other changes: 

* Instance creation now fails if the referenced `PlaybookID` is missing
* `BROADWAY_MANIFESTS_PATH` and `BROADWAY_PLAYBOOKS_PATH` env vars have been added
* Hardcoded manifest and playbook folder locations have been replaced with these env vars
* The `test` container sets these env vars set to `examples/manifests` and `examples/playbooks` respectively
* All tests that need to manipulate instances now use `PlaybookID: "helloplaybook"`
* `helloplaybook` has been added to the examples folder, it declares two vars: `bird` and `word`

### CODE REVIEW

File: deployment/manifest.go [hunk 1]
```diff
@@ -31,3 +38,33 @@ func (m *Manifest) Execute(vars map[string]string) string {
 	}
 	return b.String()
 }
+
+// SetManifestRoot ensures a folder exists, then sets ManifestRoot to that
+// folder.
+func SetManifestRoot(newRoot string) error {
```
Comment (cored) [inline deployment/manifest.go]: @therealplato Why is this related?

File: deployment/manifest.go [hunk 2]
```diff
@@ -31,3 +38,33 @@ func (m *Manifest) Execute(vars map[string]string) string {
...
+	if _, err := os.Stat(newRoot); err != nil {
+		return err
+	}
+	env.ManifestsPath = newRoot
+	return nil
+}
+
+// ManifestsPresent iterates through the Manifests and PodManifest items on a
+// task, and checks that each represents a file on disk
+func (t Task) ManifestsPresent() error {
```
Comment (cored) [inline deployment/manifest.go]: Also this?

File: env/env.go [hunk 3]
```diff
@@ -32,17 +33,41 @@ var (
...
+
+var (
+	cwd                  string
+	defaultManifestsPath string
+	defaultPlaybooksPath string
 )
 
 // LoadEnvs sets env.* variables to their OS-provided value
 func LoadEnvs() {
+	ManifestsPath = loadw("BROADWAY_MANIFESTS_PATH")
```
Comment (bobbytables) [inline env/env.go]: We probably a `sync.Once` in here so that this can't be called again.

File: server/server.go [hunk 4]
```diff
@@ -70,6 +70,7 @@ func (s *Server) Init() {
 	}
 
 	s.playbooks, err = deployment.LoadPlaybookFolder("playbooks/")
```
Comment (cored) [inline server/server.go]: @therealplato Why don't you change this to use the new environment variable?

Comment (therealplato) [inline server/server.go]: good catch

File: services/instance.go [hunk 5]
```diff
@@ -19,17 +20,57 @@ func NewInstanceService(s store.Store) *InstanceService {
 	return &InstanceService{repo: instance.NewRepo(s)}
 }
 
+// PlaybookNotFound indicates a problem due to Broadway not knowing about a
+// playbook
+type PlaybookNotFound struct {
```
Comment (cored) [inline services/instance.go]: I like this.

File: services/instance.go [hunk 6]
```diff
@@ -19,17 +20,57 @@ func NewInstanceService(s store.Store) *InstanceService {
...
+
+func (e *InvalidVar) Error() string {
+	return fmt.Sprintf("Playbook %s does not declare a var named %s\n", e.playbookID, e.key)
+}
+
 // Create a new instance
 func (is *InstanceService) Create(i *instance.Instance) (*instance.Instance, error) {
-	glog.Info("Instance Service: Create")
+	glog.Infof("InstanceService attempting to create %s/%s...\n", i.PlaybookID, i.ID)
+	pb, ok := deployment.AllPlaybooks[i.PlaybookID]
```
Comment (cored) [inline services/instance.go]: You can switch this call to be inline. 
`if pb, ok := deploymemt..; ok { }`

File: deployment/manifest.go (hunk 1 above)
Comment (therealplato) [inline deployment/manifest.go]: i moved this out of playbook.go while refactoring, i think it can now be removed since e.g. tests can manipulate `env.ManifestsPath`

File: deployment/manifest.go (hunk 2 above)
Comment (therealplato) [inline deployment/manifest.go]: This is used in playbook validation

File: services/instance.go [hunk 7]
```diff
@@ -19,17 +20,57 @@ func NewInstanceService(s store.Store) *InstanceService {
...
+	if ok != true {
+		return nil, &PlaybookNotFound{i.PlaybookID}
+	}
+	// Set all valid vars to empty string:
+	vars := make(map[string]string)
+	for _, pv := range pb.Vars {
+		vars[pv] = ""
+	}
+	// Look for invalid vars on new instance, and set values:
```
Comment (cored) [inline services/instance.go]: What does an invalid mean in this context? Why do we need to set them?

File: services/instance.go [hunk 8]
```diff
@@ -19,17 +20,57 @@ func NewInstanceService(s store.Store) *InstanceService {
...
+	i.Vars = vars
 	err := is.repo.Save(i)
 	if err != nil {
 		return nil, err
 	}
-	err = sendCreationNotification(i)
-	if err != nil {
-		return i, err
-	}
+	/*
```
Comment (cored) [inline services/instance.go]: Remove this commented code.

File: services/instance_test.go [hunk 9]
```diff
@@ -8,47 +8,78 @@ import (
...
+	service := NewInstanceService(store)
+
+	i := &instance.Instance{PlaybookID: "vanishing-pb", ID: "gone"}
+	_, err := service.Create(i)
+
+	assert.NotNil(t, err)
+	assert.Contains(t, err.Error(), "playbook vanishing-pb is missing")
+}
+
+func TestCreateInstanceWithIncorrectVars(t *testing.T) {
```
Comment (cored) [inline services/instance_test.go]: Can you add this or remove it?

File: services/instance_test.go [hunk 10]
```diff
@@ -8,47 +8,78 @@ import (
...
+	//todo: set the wrong vars
+}
+
+/*
+func TestCreateInstanceNotification(t *testing.T) {
```
Comment (cored) [inline services/instance_test.go]: Remove this commented code. Please.

File: services/instance.go (hunk 8 above)
Comment (therealplato) [inline services/instance.go]: @macat actually if you have time can you uncomment this, and the associated test, and see if you can reproduce my `net` error in sending notification?

File: services/instance_test.go (hunk 9 above)
Comment (therealplato) [inline services/instance_test.go]: yep will add these tests

File: services/instance_test.go [hunk 11]
```diff
@@ -8,47 +8,78 @@ import (
...
-	assert.Equal(t, "test", createdInstance.PlaybookID)
-	assert.Equal(t, instance.StatusNew, createdInstance.Status)
+
 	assert.Contains(t, nt.requestBody, "created")
+	assert.Contains(t, nt.requestBody, "helloplaybook")
+	assert.Contains(t, nt.requestBody, "TestCreateInstanceNotification")
+}
+*/
+
+func TestCreateInstance(t *testing.T) {
```
Comment (cored) [inline services/instance_test.go]: I think we can use the test table idea for this method.

File: services/instance_test.go [hunk 12]
```diff
@@ -8,47 +8,78 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
-func TestCreateInstance(t *testing.T) {
+func TestCreateInstanceFromMissingPlaybook(t *testing.T) {
```
Comment (cored) [inline services/instance_test.go]: @therealplato Is there a test for the create that stress the code for 'invalid variables' ? I can't see it in here.

File: services/instance.go (hunk 6 above)
Comment (therealplato) [inline services/instance.go]: if i did that, wouldn't pb be out of scope after the if block? i need to use it after the if block

File: services/slack_test.go [hunk 13]
```diff
@@ -26,56 +21,56 @@ func TestSetvarExecute(t *testing.T) {
 		E            error
 	}{
 		{
-			"When an instance and playbook defines a variable",
-			"setvar foo bar var1=val1",
-			&instance.Instance{PlaybookID: "foo", ID: "bar", Vars: map[string]string{"var1": "val2"}},
-			map[string]*deployment.Playbook{"foo": &deployment.Playbook{ID: "foo", Vars: []string{"var1"}}},
-			map[string]string{"var1": "val1"},
-			"Instance foo bar updated it's variables",
+			"When an instance sets a new variable value from the playbook",
```
Comment (cored) [inline services/slack_test.go]: This is test is what suggest me that we need to inject the playbooks into the `.Create` method in the instance service as a parameter. I know we discuss this but the refactoring could be part of the same PR. What do you think?

File: services/instance.go (hunk 6 above)
Comment (cored) [inline services/instance.go]: @therealplato Try it :-)

**Comment by poopoothegorilla** [comment]
+1 looks good :)

**Comment by macat** [comment]
+1
